### PR TITLE
Cleanup split logic

### PR DESCRIPTION
**Pull request by richardpringle**
Although it's still relatively complex

### CODE REVIEW

File: firewood/src/merkle.rs [hunk 1]
```diff
@@ -161,179 +164,213 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
...
-                        child_encoded: None,
-                    })))?
-                    .as_ptr()
-                } else {
-                    branch_ptr
+        let node_to_split_address = node_to_split.as_ptr();
+        let split_index = insert_path
+            .iter()
+            .zip(n_path.iter())
+            .position(|(a, b)| a != b);
```
Comment (rkuris) [inline firewood/src/merkle.rs:179]: 🎸

File: firewood/src/merkle.rs [hunk 2]
```diff
@@ -161,179 +164,213 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
...
+                branch_address
             }
         };
+
         // observation:
         // - leaf/extension node can only be the child of a branch node
         // - branch node can only be the child of a branch/extension node
-        set_parent(new_chd, parents);
+        // ^^^ I think a leaf can end up being the child of an extension node
+        // ^^^ maybe just on delete though? I'm not sure, deleting extensions anyway
```
Comment (richardpringle) [inline firewood/src/merkle.rs:371]:
```suggestion
        // ^^^ maybe just on delete though? I'm not sure, removing extension-nodes anyway
```

File: firewood/src/merkle/node.rs [hunk 3]
```diff
@@ -409,6 +401,24 @@ impl NodeType {
             )))),
         }
     }
+
+    pub fn encode<S: ShaleStore<Node>>(&self, store: &S) -> Vec<u8> {
```
Comment (richardpringle) [inline firewood/src/merkle/node.rs:405]: Keeping static methods defined at the top of the impl block

File: firewood/src/merkle/node.rs [hunk 4]
```diff
@@ -409,6 +401,24 @@ impl NodeType {
...
+    }
+
+    pub fn path_mut(&mut self) -> Option<&mut PartialPath> {
+        let path = match self {
+            NodeType::Branch(_) => return None,
+            NodeType::Leaf(node) => &mut node.0,
+            NodeType::Extension(node) => &mut node.path,
+        };
+
+        Some(path)
```
Comment (richardpringle) [inline firewood/src/merkle/node.rs:420]:
```suggestion
        path.into()
```

File: firewood/src/merkle.rs [hunk 5]
```diff
@@ -161,179 +164,213 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
...
-                                                    Some(Data(val));
-                                                b.rehash()
-                                            });
+                _ => node_to_split_address,
+            };
+
+            chd[n_path[idx] as usize] = Some(address);
+
+            // don't think I need this
+            // drop(node_to_split);
```
Comment (rkuris) [inline firewood/src/merkle.rs:214]: OK so let's nuke it then

File: firewood/src/merkle.rs [hunk 6]
```diff
@@ -161,179 +164,213 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
...
-                        child: branch_ptr,
-                        child_encoded: None,
-                    })))?
-                    .as_ptr()
-                } else {
-                    branch_ptr
-                }
+
+            // I donn't think I need this
+            // drop(node_to_split);
```
Comment (rkuris) [inline firewood/src/merkle.rs:340]: Typo, and remove it

**Review by rkuris — APPROVED** [review]

**Review by xinifinity — APPROVED** [review]

**Pull request merged**